Add email to Identifier and new layer for API access

camposer opened the pull request.

wallrony committed:
Add command text regex and get member details func (strongdm/accessbot#151)

Co-authored-by: vassalo <vassalo@users.noreply.github.com>

wallrony committed:
Add user details, conversation and activity methods (strongdm/accessbot#151)

wallrony committed:
Get member info when message sent by DM (strongdm/accessbot#151)

Co-authored-by: vassalo <vassalo@users.noreply.github.com>

wallrony committed:
Add ms teams webclient and tests base (strongdm/accessbot#151)

Co-authored-by: vassalo <vassalo@users.noreply.github.com>

vassalo committed:
Add botframework tests (strongdm/accessbot#151)

Co-authored-by: wallrony <wallrony@users.noreply.github.com>

vassalo committed:
Improve botframework tests (strongdm/accessbot#151)

Co-authored-by: wallrony <wallrony@users.noreply.github.com>

camposer committed:
Merge pull request #1 from strongdm/feat/accessbot-integration

Feat/accessbot integration

vasilcovsky commented on ms_teams_webclient.py:
```diff
@@ -0,0 +1,153 @@
+            headers=self.__get_default_headers()
+        )
+        if response.status_code == 404:
+            raise MemberNotFound(f"member not found using {id} id")
```
{id} is not defined

vasilcovsky commented on ms_teams_webclient.py:
```diff
@@ -0,0 +1,153 @@
+            f'{service_url}/v3/conversations/{team_id}/members/{email}',
+            headers=self.__get_default_headers()
+        )
+        if response.status_code == 404:
+            raise MemberNotFound(f"member not found using {identifier.email} email")
```
I believe that raise_for_status is much better than custom exception because it preserves details from service's response
```suggestion
r.raise_for_status()
```

camposer replied:
We added `raise_for_status` in order to keep the stack trace, but we still have a custom Exception so we can better treat the error at the bot side.

vasilcovsky commented on botframework.py:
```diff
@@ -255,16 +226,22 @@ def get_botframework():
         def post_botframework():
             req = request.json
+
+            req['text'] = unicodedata.normalize("NFKD", req['text']).encode('ascii', 'ignore').decode('UTF-8')
```
not clear why this is needed

camposer replied:
You have a very valid point. We added a new function with a self-describing name and some documentation. Thanks for the feedback

vasilcovsky commented on ms_teams_webclient.py:
```diff
@@ -0,0 +1,153 @@
+    def __create_conversation(self, member_id, extras):
+        body = {
+            "bot": {
+                "id": f"28:{self.__app_id}"
```
what is 28?

camposer replied:
Good point! This prefix is needed for Azure Bot implementations. We added a new constant for making it easier to understand.

vasilcovsky commented on test_botframework.py:
```diff
@@ -0,0 +1,76 @@
+default_reply = {
+    'type': 'message',
+    'text': 'My text message'
+}
```
lines above should be fixture(s)

camposer replied:
Honestly we don't think fixtures are really needed here. As one downside, you'd end up with very long functions with a bunch of parameters. As in many other programing languages, we think that just a variable would be enough here.

vasilcovsky commented on test_botframework.py:
```diff
@@ -0,0 +1,76 @@
+    'text': 'My text message'
+}
+
+class Test_send_message:
```
since you are using `pytest`, could you please remove `class` and move test to module's level?

camposer replied:
We can definitively remove classes, but we usually prefer to use groupings (tests suites), as you would use with any BDD framework (e.g. [Jasmine](https://jasmine.github.io/)). So the class, is just a shortcut for wrapping all the tests related to a method. Without classes, you would need to have very long names for your methods and the file tends to become hard to read.

vasilcovsky commented:
Thanks for PR! What kind of new functionality does it bring?

camposer commented:
Thank you @vasilcovsky !

The PR adds the possibility of sending-receiving messages to/from a Team and at the same time adds email info to the Identifier. We needed it for developing the MS Teams version of [AccessBot](https://github.com/strongdm/accessbot)

Thanks to your contribution, we made it work with very few changes!

We should update the PR during the week and comment you here :-)

wallrony committed:
Improve code readability (strongdm/accessbot#151)

- Describe utf8 normalization process;
- Add stacktrace in webclient exceptions;

Co-authored-by: vassalo <vassalo@users.noreply.github.com>

Commit:
Improve webclient and tests (strongdm/accessbot#151)

camposer committed:
Merge pull request #2 from strongdm/feat/accessbot-integration

Improve code readability

camposer commented:
Thanks again for your contribution to the community. Looking forward to continue the alliance. Keep the good work!

wallrony committed:
Add extra info in from identifier

wallrony committed:
Add reply functionality (strongdm/accessbot#186)

Co-authored-by: vassalo <vassalo@users.noreply.github.com>

camposer committed:
Merge pull request #4 from strongdm/feat/reply-message

Add reply functionality

camposer committed:
Merge pull request #3 from strongdm/feat/add-extras-in-frm

Add extra info in from identifier

wallrony committed:
Add tests workflow action

Co-authored-by: vassalo <vassalo@users.noreply.github.com>

wallrony committed:
Update tests workflow branch

Co-authored-by: vassalo <vassalo@users.noreply.github.com>

wallrony committed:
Merge pull request #5 from strongdm/feat/add-tests-ci

Add tests workflow action

wallrony committed:
Add microsoft graph webclient for retrieving user details

Co-authored-by: vassalo <vassalo@users.noreply.github.com>

wallrony committed:
Fix webclient reference on botframework tests

Co-authored-by: vassalo <vassalo@users.noreply.github.com>

wallrony committed:
Add credentials config validation (strongdm/accessbot#295)

Co-authored-by: vassalo <vassalo@users.noreply.github.com>

vassalo committed:
Fix credentials error on tests (strongdm/accessbot#295)

Co-authored-by: wallrony <wallrony@users.noreply.github.com>

camposer committed:
Merge pull request #6 from strongdm/feat/add-graph-webclient

Add graph webclient

wallrony committed:
Add get team and channel api request methods

Co-authored-by: vassalo <vassalo@users.noreply.github.com>

wallrony committed:
Add validation on build_identifier when parsing a channel name

Co-authored-by: vassalo <vassalo@users.noreply.github.com>

wallrony committed:
Improve identifier construction and isolate webclient methods usage (strongdm/accessbot#297)

vassalo committed:
Improve azure credentials/authorization exceptions and fix ADMINS_CHANNEL_ELEVATE (strongdm/accessbot#297)

Co-authored-by: wallrony <wallrony@users.noreply.github.com>

vassalo committed:
Fix tests and add email_validator to requirements/common

Co-authored-by: wallrony <wallrony@users.noreply.github.com>

wallrony committed:
Add tests for send and build_identifier methods (strongdm/accessbot#297)

Co-authored-by: vassalo <vassalo@users.noreply.github.com>

wallrony committed:
Add tests for get channel by id method (#297)

Co-authored-by: vassalo <vassalo@users.noreply.github.com>

camposer committed:
Merge pull request #7 from strongdm/feat/add_get_team_and_channel_info

Improve build identifier and azure credentials/authorization validations

vassalo committed:
Fix ADMINS_CHANNEL_ELEVATE on MS Teams when using the default channel (strongdm/accessbot#297)

Co-authored-by: wallrony <wallrony@users.noreply.github.com>

camposer committed:
Merge pull request #8 from strongdm/feat/add_get_team_and_channel_info

Remove default channel name
